### ISSUE TITLE
PAM sessions opened by PAMName= are not properly closed

### ISSUE THREAD

**Issue by fbuihuu** [systemd/systemd#14029]
That's due to the fact that user@.service uses `KilledMode=mixed` which means that PAM session handler process (sd-pam) will be SIGKILLed as soon as user instance will exit. This gives to sd-pam no chances to close the PAM session properly.

The mixed mode has been introduced by 58ea275a68cd242ad60161bcb7582614d1d89f13 and according the commit message this was done on purpose to fix "some" races. So I'm reluctant to change the mode.

**Comment by poettering** [comment]
Hmm, this reminds me of #1350...

(Where did you notice this btw? It's probably wise if people avoid the session exit handler anyway if they want somethong robust... i.e. do it like logind does: open an fd in the session enter handler, and wait in the bg until you get an EOF on it, so that you can cover any kind of session termination, clean or not)

**Comment by jmoellers** [comment]
For one ... if pam_mount is used to eg auto-mount encrypted partitions when a user logs in.
On SUSE Linux, pam_mount.so is added to /etc/pam.d/common-session (and common-auth, but that's not causing problems) which is included by just about every PAM configuration file, including systemd-user. So systemd opens a (PAM-) session (which causes pam_mount to mount the user's $HOME) and then the login process (eg sshd) opens another (PAM-) session which doesn't mount the user's $HOME as it is already mounted.
When the user logs out, sshd closes the (PAM-) session which doesn't do anything to the user's $HOME due to a reference count, which pam_mount keeps, not being down to 0. Later, sd-pam also closes the (PAM-) session and then pam_mount is invoked again and the reference count drops to 0  and pam_mount would like to umount the user's $HOME as the reference count now drops to 0.
But since sd-pam now runs with the user's privileges, this fails as the user is not allowed to umount his own $HOME.

**Comment by jmoellers** [comment]
We see this when pam_mount.so is added to eg auto-mount a user's encrypted $HOME partition.
Everything is fine until sd-pam should close the (PAM-) session which would finally unmount $HOME but is killed before that.

**Comment by fbuihuu** [comment]
> Hmm, this reminds me of #1350...

But I think this issue is a different one. This one is about the fact that sd-pam is SIGKILLed whereas it supposed to receive SIGTERM and close the PAM session properly. It's its main purpose after all...

> (Where did you notice this btw?

Josef described the case where we noticed that sd-pam wasn't exiting properly. And this case is indeed the same problem as the one described by #1350, i.e. some privileged actions might be involved by pam_close_session().

**Comment by jmoellers** [comment]
I think these are two different aspects:
1) sd-pam has dropped privileges so it may not be able to properly close the (PAM-) sesson (eg when closing the session involves umounting something)
2) before sd-pam can fail to do some privileged operation, it is killed ;-)

**Comment by poettering** [comment]
as mentioned, i would really advise against relying on pam session exit handlers for this kind of stuff if you want this robust. Instead, allocate a socket or pipe fd or so in the pam session enter handler with a small daemon on the other end. Then just close the pipe on the exit handler. benefit of this is that if for some reason the session is killed abnormally (like in this case) you still can clean up things properly. it's how logind does it for example.

i.e. yes we should probably be friendlier to pam exit handlers, but relying on them exclusively is not a good idea anyway...

**Comment by fbuihuu** [comment]
Well first of all pam session exit handlers are the common pattern for PAM modules to clean their stuff up so I don't think we can just ignore this fact and ask for the PAM modules to be rewritten.

Moreover even if they would be rewritten like you described, I don't think that would help because the small daemon would be also killed by PID1 with a SIGKILL like sd-pam is so that wouldn't help.

Again the problem here is that sd-pam is killed with SIGKILL as soon as 'systemd --user' exits.

**Comment by poettering** [comment]
well, as mentioned, I think we should be friendler to pam exit handlers, but all I am saying: write your stuff robust, don't rely on the exit handlers, it just means you leave stuff around that might never be cleaned up when something goes wrong.

**Comment by poettering** [comment]
> Moreover even if they would be rewritten like you described, I don't think that would help because the small daemon would be also killed by PID1 with a SIGKILL like sd-pam is so that wouldn't help.

huh? it would of course live in its own little unit, maybe socket activated even.  forking off additional long-running proceses from a PAM session handlers is just evil. don't even think about it...

**Comment by poettering** [comment]
> Again the problem here is that sd-pam is killed with SIGKILL as soon as 'systemd --user' exits.

fixing this properly is hard. We need to rearrange how the PAM stuff works, entirely. for this case and for others. i.e. we need to make sure PID1 manages the process explicitly like it manages the main and control processes. moreover the PAM stub is currently a CoW trap, since it is forked off PID 1 and sticks around forever. Hence to make this clean we'd need to substantially beef up the forking logic in PID 1, probably serialize all properties to execute in json, then introduce one level of execve() towards a second helper binary which the deserializes it and continues with that, when PAM is required (to turn off the CoW effect), then actively tell PID 1 about our little PAM process, so that PID 1 tracks it explicitly and can give it time to do what it wants, so that we don#t have to rely on PR_PDEATHSIG and pid1's cgroup killing anymore for clean-up.

Hence, don#t hold your breath... my recommendation would be: fix your stuff so that it doesn#t need this, it's cleaner and more robust in all contexts since you don't have to rely on the PAM exit hook correctly working anymore.

**Comment by fbuihuu** [comment]
Or maybe just don't reap the main process until sd-pam exits ?

That said this suggests that the PAMName= option might have been introduced too prematurely and we should have let services that need to open a PAM session do that from their main process.

BTW I agree that the PAM exit session handlers seems borked but unfortunately we need to support the normal case.

**Comment by poettering** [comment]
> Or maybe just don't reap the main process until sd-pam exits ?

Currently PID 1 doesn't track sd-pam specifically, it's like any other process for it. That's why I am saying above: we need to start tracking it explicitly before we can change anything here.

> That said this suggests that the PAMName= option might have been introduced too prematurely and we should have let services that need to open a PAM session do that from their main process.

Well PAMName= is like 7 years old or so. The assumption was that stuff htat actually has resources to manage (and mounts are resources) wouldn't rely on the PAM exit hook exclusively, and so far we got away with it just fine... It's the first time this comes up...

> BTW I agree that the PAM exit session handlers seems borked but unfortunately we need to support the normal case.

Well, you are welcome to work on a patch...

(BTW: related to this: I was intending to add ExecMonitor= one of those days, which would allow definition of processes that are spawned as part of a service, after the service is up and which when they die in any form cause the service to shutdown. For that we need to track more than just the control and main process anyway. So I figure revamping the PAM support could also add ExecMonitor= support at the same time, it should be easy then. Ultimately for each process we should probably watch: 1x main pid + 1x control pid + 1x PAM session stub process pid for main + 1x PAM session stub process pid for control + N x ExecMonitor= processes...)

**Comment by fbuihuu** [comment]
> Well PAMName= is like 7 years old or so. The assumption was that stuff htat actually has resources to manage (and mounts are resources) wouldn't rely on the PAM exit hook exclusively, and so far we got away with it just fine... It's the first time this comes up...
> 

pam_mount is affected by this and https://bugs.freedesktop.org/show_bug.cgi?id=72759 might be related too.

Anyway any PAM modules that rely on pam_close_session() to clean their stuff up is just broken currently even for the normal path.

> > BTW I agree that the PAM exit session handlers seems borked but unfortunately we need to support the normal case.
> 
> Well, you are welcome to work on a patch...

Before working on a patch we need to discuss on possible solutions first since as you said it's not really trivial, right ?

> (BTW: related to this: I was intending to add ExecMonitor= one of those days, which would allow 

Noted I'll try to think about it when time will permit.

**Comment by jmoellers** [comment]
Setting
SendSIGKILL=no
in user@.service allowed pam_close_session to finish.
It's not the solution, but it may help getting forward.

**Comment by fbuihuu** [comment]
> Well PAMName= is like 7 years old or so. The assumption was that stuff htat actually has resources to manage (and mounts are resources) wouldn't rely on the PAM exit hook exclusively, and so far we got away with it just fine... It's the first time this comes up...

Ok this is definitively not true and in fact I just realized that this issue has already been opened more than 1 year ago: https://github.com/systemd/systemd/issues/8598

**Comment by poettering** [comment]
Let's close this as duplicate of #8598
